// @sap/grunt-sapui5-bestpractice-build/tasks/updateNeoApp.js
var path = require('path');

var _NEO_JSON_FILE_NAME = "neo-app.json";

function updateNeoApp(grunt, srcFolder, appFolder, distDir) {
	var sNeoAppFile = path.join(srcFolder, _NEO_JSON_FILE_NAME);
    var bNeoAppExists = grunt.file.exists(sNeoAppFile);

    if (bNeoAppExists) {
        try {
            var oNeoAppJsonContent = grunt.file.readJSON(sNeoAppFile);
        }
        catch (error) {
            throw new Error("devxUpdateNeoApp - Could not read neo-app.json, bad format?");
        }
        updateNeoAppFile(grunt, srcFolder, appFolder, distDir, oNeoAppJsonContent);
   }
    else {
        throw new Error("devxUpdateNeoApp - No neo-app.json file found");
    }
};

function updateNeoAppFile(grunt, srcFolder, appFolder, distDir, oNeoAppJsonContent){
	var sSourceFolderRelativePath = "/" + path.relative(srcFolder, appFolder);

	var routes = oNeoAppJsonContent.routes;
	for (var index in routes) {
		//remove source folder from path
		if (routes[index].path.indexOf(sSourceFolderRelativePath) === 0) {
			routes[index].path = routes[index].path.substring(sSourceFolderRelativePath.length);
		}
	}

	//Remove source folder in welcome file
	var welcomeFile = oNeoAppJsonContent.welcomeFile;
	if (welcomeFile) {
		if (welcomeFile.indexOf(sSourceFolderRelativePath) === 0) {
			oNeoAppJsonContent.welcomeFile = welcomeFile.substring(sSourceFolderRelativePath.length);
		}
	}

	var newNeoappFile = path.join(distDir, _NEO_JSON_FILE_NAME);
	try {
		grunt.file.write(newNeoappFile, JSON.stringify(oNeoAppJsonContent, null, 2));
	}
	catch(error){
		throw new Error("devxUpdateNeoApp - failed to write updated neo-app.json file");
	}
}

function validateTaskOptions(grunt, task, taskOptions) {
    if (!taskOptions.src) {
        grunt.log.error(task.name + ' missing src configuration');
        return false;
    }
    if (!taskOptions.dest) {
        grunt.log.error(task.name + ' missing dest configuration');
        return false;
    }
    if (!taskOptions.app) {
        grunt.log.error(task.name + ' missing app configuration');
        return false;
    }
    return true;
}

function getGruntConfiguredOptions(grunt) {
    return {
        src: grunt.config.get('dir.projectRoot'),
        app: grunt.config.get('dir.appFolder'),
        dest: grunt.config.get('dir.dist')
    }
}

module.exports = function(grunt) {
	grunt.registerTask('devxUpdateNeoApp', 'Update neo-app for build', function () {
        //get tasks configured options or set default if not exists
        var originOptions = this.options(getGruntConfiguredOptions(grunt));
        var _options = originOptions;
		try {
            if (validateTaskOptions(grunt, this, _options)) {
                updateNeoApp(grunt, _options.src, _options.app, _options.dest);
            }
		} catch(error) {
			grunt.log.error("Failed to update neo-app.json file: " + error.message);
		}
	});
}
